Image registries management (#18193)
canonical/lxd: doc/howto/image_registries.md

---
myst:
  html_meta:
    description: Create, view, configure, and delete LXD image registries.
---

(howto-image-registries)=
# How to manage image registries

> Image registries are global read-only sources of images. They provide a way to define image sources (like SimpleStreams or other LXD servers) that are accessible to all members of the cluster and are managed centrally.

Author: Updated.

> Once created, an image registry can be used to download images for LXD.

Author: Updated.

(howto-image-registries-create)=
## Create an image registry

> How to add an image registry depends on the protocol that the source uses.

Author: Updated.

`````{tabs}
````{group-tab} CLI

To add a registry using the `lxd` protocol (another LXD server), you link to it through a {ref}`cluster link <howto-cluster-links-manage>`.
Use a public cluster link for a public LXD server, or an authenticated cluster link for a private one:

    lxc image registry create <registry_name> --protocol=lxd cluster=<cluster_link_name> source_project=<project>

> 

Author: Updated.

To add a registry using the `simplestreams` protocol, run:

>     lxc image registry create <registry_name> --protocol=simplestreams url=<URL>

Author: Yes, the `url` key is required for `simplestreams` registries.

>     lxc image registry create <registry_name> --protocol=simplestreams url=<URL>

Author: Updated.

    lxc query --request POST /1.0/image-registries --data '{"name": "my-registry", "protocol": "lxd", "config": {"cluster": "my-cluster-link", "source_project": "default"}}'

See [`POST /1.0/image-registries`](swagger:/image-registries/image_registries_post) for more information.

````
`````

(howto-image-registries-view)=
## View image registries

`````{tabs}
````{group-tab} CLI

To list all configured image registries, run:

    lxc image registry list

To view the configuration of a specific image registry, run:

    lxc image registry show <registry_name>

````
````{group-tab} API

To list all image registries, send the following request:

    lxc query --request GET /1.0/image-registries

To view the configuration of a specific image registry, send the following request:

    lxc query --request GET /1.0/image-registries/<name>

See [`GET /1.0/image-registries`](swagger:/image-registries/image_registries_get) and [`GET /1.0/image-registries/{name}`](swagger:/image-registries/{name}/image_registry_get) for more information.

````
`````

(howto-image-registries-configure)=
## Configure an image registry

> You can update the configuration for an image registry by editing it or by setting specific properties.

Author: Updated.

`````{tabs}
````{group-tab} CLI

To edit an image registry in your default text editor, run:

    lxc image registry edit <registry_name>

To set a specific configuration option, run:

    lxc image registry set <registry_name> <key> <value>

For example, to update the source project:

    lxc image registry set my-registry source_project foo

````
````{group-tab} API

To update an image registry, send a `PUT` or `PATCH` request to the `/1.0/image-registries/<name>` endpoint.

    lxc query --request PATCH /1.0/image-registries/my-registry --data '{"description": "New description"}'

See [`PUT /1.0/image-registries/{name}`](swagger:/image-registries/{name}/image_registry_put) and [`PATCH /1.0/image-registries/{name}`](swagger:/image-registries/{name}/image_registry_patch) for more information.

````
`````

Image registries have the following configuration options:

% Include content from [../metadata.txt](../metadata.txt)
```{include} ../metadata.txt
    :start-after: <!-- config group image-registry-image-registry-conf start -->
    :end-before: <!-- config group image-registry-image-registry-conf end -->

> ```

Author: Updated.

# How to manage image registries

Image registries are global read-only sources of images. They provide a way to define image sources (like SimpleStreams or other LXD servers) that are accessible to all members of the cluster and are managed centrally.
Once created, an image registry can be used to download images for LXD.

(howto-image-registries-create)=
## Create an image registry

How to add an image registry depends on the protocol that the source uses.

`````{tabs}
````{group-tab} CLI

To add a public registry using the `lxd` protocol (another LXD server), run:

    lxc image registry create <registry_name> --protocol=lxd url=<URL> source_project=default

If the source LXD server is private and linked via a {ref}`cluster link <howto-cluster-links-manage>`, you can specify the cluster link name:

    lxc image registry create <registry_name> --protocol=lxd cluster=<cluster_link_name> source_project=<project>

To add a registry using the `simplestreams` protocol, run:

    lxc image registry create <registry_name> --protocol=simplestreams url=<URL>

````
````{group-tab} API

To add an image registry, send a `POST` request to the `/1.0/image-registries` endpoint.

>     lxc query --request POST /1.0/image-registries --data '{"name": "my-registry", "protocol": "lxd", "config": {"url": "https://192.0.2.10:8443", "source_project": "default"}}'

Reviewer: In this case `https://192.0.2.10:8443` likely has a self-signed TLS cert so I'm wondering if the expected fingerprint also be relayed.

AFAIK, the current way of pulling images from the client comes with TLS verification usually anchored in the fingerprint match, sometimes by walking the trust up to a trusted CA. Would we be losing this with image registries?

Author: This is currently a temporary workaround. Basically the image registry fetches the self-signed TLS cert each time and trusts it as is.

Once `unauthenticated` cluster links PR is merged, I'll update this PR to use them instead for public LXD registries. As a result, the TLS cert will be handled by the cluster link and the user would need to provide the `cluster` field instead of the `url` when creating a public LXD registry.

> AFAIK, the current way of pulling images from the client comes with TLS verification usually anchored in the fingerprint match, sometimes by walking the trust up to a trusted CA. Would we be losing this with image registries?

The `unauthenticated` cluster links should preserve this behavior as far as I understand https://github.com/canonical/lxd/pull/18197.

Reviewer: Thanks for the pointer, I headed over to #18197 to ask for more information about this `unauthenticated` cluster link.

If we ignore that other PR for now, if we blindly trust a TLS cert without any verification, I think it would need to be explicitly done with a clear flag.

## View image registries

`````{tabs}
````{group-tab} CLI

To list all configured image registries, run:

    lxc image registry list

To view the configuration of a specific image registry, run:

    lxc image registry show <registry_name>

````
````{group-tab} API

To list all image registries, send the following request:

    lxc query --request GET /1.0/image-registries

To view the configuration of a specific image registry, send the following request:

    lxc query --request GET /1.0/image-registries/<name>

See [`GET /1.0/image-registries`](swagger:/image-registries/image_registries_get) and [`GET /1.0/image-registries/{name}`](swagger:/image-registries/{name}/image_registry_get) for more information.

````
`````

(howto-image-registries-configure)=
## Configure an image registry

You can update the configuration for an image registry by editing it or by setting specific properties.

`````{tabs}
````{group-tab} CLI

To edit an image registry in your default text editor, run:

    lxc image registry edit <registry_name>

To set a specific configuration option, run:

    lxc image registry set <registry_name> <key> <value>

For example, to update the description:

>     lxc image registry set my-registry cluster lxd02

Author: Fixed.

---
myst:
  html_meta:
    description: Create, view, configure, and delete LXD image registries.
---

(howto-image-registries)=
# How to manage image registries

> Image registries are global read-only sources of images.

Reviewer 2: Quibbling a bit here, but this makes it sound like the "image registry" is the actual source of the image, as if "image registry" = "image source" (i.e. the other LXD server). Can you reword so that it's clear what the image registry is on the LXD side? It seems that the image registry is a kind of list of information about the images source and the images available there.

Rather than saying what an image registry *is* (since "registry" provides a decent hint), maybe you can describe it in terms of what it *does*? e.g. Image registries store information about global, read-only sources of images. Does that cover the scope of what the image registry does? (My example is perhaps missing the "manage images" element, though.)

Author: > Image registries store information about global, read-only sources of images. Does that cover the scope of what the image registry does?

Yes, I think so.

The "manage images" part is being done on the LXD side without the help of image registries (they only act as a source from where to download an image, and once downloaded, LXD takes care of managing the image locally).

Author: Updated.

You can use image registries to define and centrally manage image sources (like SimpleStreams or other LXD servers) that are accessible to all members of the cluster.
Once you create an image registry, you can use it to download images for LXD.

(howto-image-registries-create)=
## Create an image registry

> How to add an image registry depends on the protocol used by the source.

Reviewer 2: The section title and the command uses "create." Should this be consistent in the description? When you say "add", it sounds as if the registry already exists somewhere "out there" (e.g. on the images source), and you're *adding* that object to LXD.

Reviewer 2: Also, I'm wondering if we can reword, to not begin the sentence with "How". How about something like:

"The requirements for creating a new image registry depend on the protocol used by the images source."

Author: Makes sense, I'll update.

Author: Updated.

> To add a registry using the `lxd` protocol, you must connect to the remote LXD server through a {ref}`cluster link <howto-cluster-links-create>`.

Reviewer 2: This sounds as if you create the registry through the `lxd` protocol, but it seems as if the protocol is what is used on the images source (server?) side. Is that correct, and does it require clarification? If so, maybe something like ... "To add/create a registry for an images source/server that uses the `lxd` protocol..." (simplestreams would require a similar update).

Author: Yes, the protocol is what is used on the images source side. I'll rephrase the sentence.

Author: Updated.

> Use a public cluster link for a public LXD server, or an authenticated cluster link for a private one.

Reviewer 2: Current cluster link types are `bidirectional`, `unidirectional`, or `public`. "Authenticated" seems to suggest either `bidirectional` or `unidirectional`. Would users typically use one or both of these types of cluster links in this scenario? We'd probably want to keep the word "authenticated" here, though; it's a nice contrast against "public."

How about the following (and remove the word bidirectional or unidirectional, if only one type typically applies to this use case):

```suggestion
Use a public cluster link for a public LXD server, or a bidirectional or unidirectional cluster link to authenticate to a private server.
```

Author: Ideally, the cluster link types used for image registries should be either `unidirectional` or `public`. A `bidirectional` link can also be used instead of `unidirectional`, but it is unnecessary for the image registry use case. It might be useful if the user also has some additional use cases in mind (e.g. setting up a replicator to the same server that the image registry points to).

But overall I agree with your suggestion. I'll update the wording.

Author: Updated.
